common/timedcache/timedcache.go

```diff
@@ -25,19 +25,49 @@ func (te *timedEntry) expired() bool {
 // to expire at exactly the ttl time. The expiration mechanism is 'lazy', and
 // will only remove expired objects at next access.
 type TimedCache struct {
-	ttl   int64     // Time (in seconds) each entry is allowed to live for
-	cache lru.Cache // Underlying size-limited LRU cache
+	ttl   int64
+	cache *lru.Cache
 	lock  sync.RWMutex
+
+	evictedKeys, evictedVals []interface{}
+	onEvictedCB              func(k, v interface{})
 }
 
 // New creates a new cache with a given size and ttl. TTL defines the time in
 // seconds an entry shall live, before being expired.
 func New(size int, ttl int) (*TimedCache, error) {
-	cache, err := lru.New(size)
+	return NewWithEvict(size, ttl, nil)
+}
+
+// NewWithEvict constructs a fixed size cache with the given ttl & eviction
+// callback.
+func NewWithEvict(size int, ttl int, onEvicted func(key, value interface{})) (*TimedCache, error) {
+	tc := &TimedCache{
+		ttl:         int64(ttl),
+		onEvictedCB: onEvicted,
+	}
+	if onEvicted != nil {
+		tc.initEvictBuffers()
+		onEvicted = tc.onEvictedCB
```

Comment:
Why reassign the pointer?

Reply:
This is a simple way to simultaneously signal to the `lru.NewWithEvict()` that we need a callback AND we want to trap that callback with our own `tc.onEvictedCb()`.

fwiw I didn't invent this. The lru cache does this with its underlying simple cache type and callback. I thought it was elegant and simple, so I copied.